plugin/documents.py

```diff
@@ -520,12 +520,18 @@ def _on_hover_gutter_async(self, point: int) -> None:
                 on_navigate=lambda href: self._on_navigate(href, point))
 
     def on_text_command(self, command_name: str, args: Optional[dict]) -> Optional[Tuple[str, dict]]:
+        format_on_paste = self.view.settings().get('lsp_format_on_paste', userprefs().lsp_format_on_paste)
         if command_name == "auto_complete":
             self._auto_complete_triggered_manually = True
         elif command_name == "show_scope_name" and userprefs().semantic_highlighting:
             session = self.session_async("semanticTokensProvider")
             if session:
                 return ("lsp_show_scope_name", {})
+        elif command_name == 'paste_and_indent' and format_on_paste:
+            # if `lsp_format_on_paste` is enabled and paste_and_indent is run,
+            # it is easier to find the region to format when `paste` is invoked,
+            # so we intercept the `paste_and_indent` and replace it with the `paste` command.
+            return ('paste', {})
```

Comment:
I'm not sure right now, but isn't this ViewEventListener created for all views, regardless of whether LSP is active or not? So I think this needs something like
```python
session = self.session_async(...)
if session:
    ...
```
otherwise it would always modifiy this command even when used for unrelated files. This check is also needed for the case if a server doesn't support range formatting. Also I'm unsure whether this is the best approach to do it, but I haven't thought about alternatives yet (but maybe it's fine).

Reply:
I would really prefer that we let ST format things and then do formatting on top of that but I won't demand that as I know it would be rather hard to handle it in a clean way.

Reply:
I will look into this, thanks

Reply:
I'll just say right away that if the implementation would have to get too complex or hacky then I'd prefer that you don't waste your time.

Reply:
> oh, I suppose you were referring to the actual issue that @jwortmann raised.

Yes, I was referring to the message from @jwortmann. I didn't saw your message at the time I was writing the comment.

Reply:
[f654bc8](https://github.com/sublimelsp/LSP/pull/2397/commits/f654bc813a0aa7b222efe461934759e3ca794a94)